File: TypeScript/Algorithm/Algorithm I/567. Permutation in String.ts
/**
 * Checks if a substring of s2 contains a permutation of s1.
 * @param {string} s1 - The first string to be checked for permutation.
 * @param {string} s2 - The second string to be checked for the presence of a permutation of s1.
 * @returns {boolean} Returns true if s2 contains a permutation of s1, otherwise false.
 */
function checkInclusion(s1: string, s2: string): boolean {
  const S1_LENGTH: number = s1.length;
  const S2_LENGTH: number = s2.length;

  const stringOneContent: number[] = new Array(26).fill(0);
  const stringTwoContent: number[] = new Array(26).fill(0);

  for (let index: number = 0; index < S1_LENGTH; index++) {
    stringOneContent[s1.charCodeAt(index) - 97]++;
    stringTwoContent[s2.charCodeAt(index) - 97]++;
  }

  for (let index: number = S1_LENGTH; index < S2_LENGTH; index++) {
    if (stringOneContent.every((count, i) => count === stringTwoContent[i]))
      return true;

    const charToAdd: number = s2.charCodeAt(index) - 97;
    const charToRemove: number = s2.charCodeAt(index - S1_LENGTH) - 97;

    stringTwoContent[charToAdd]++;
    stringTwoContent[charToRemove]--;
  }

  return stringOneContent.every((count, i) => count === stringTwoContent[i]);
}

File: Java/Algorithm/Algorithm I/567. Permutation in String.java
class Solution {

  /**
   * Checks if a permutation of string s1 is contained within string s2.
   *
   * @param s1 The string to be checked for permutation.
   * @param s2 The string where the permutation of s1 is to be checked.
   * @return True if a permutation of s1 is found within s2, otherwise false.
   */
  public boolean checkInclusion(String s1, String s2) {
    int S1_LENGTH = s1.length();
    int S2_LENGTH = s2.length();

    if (S1_LENGTH > S2_LENGTH) {
      return false;
    }

    int[] stringOneContent = new int[26];
    int[] stringTwoContent = new int[26];

    for (int index = 0; index < S1_LENGTH; index++) {
      stringOneContent[s1.charAt(index) - 'a']++;
      stringTwoContent[s2.charAt(index) - 'a']++;
    }

    for (int index = S1_LENGTH; index < S2_LENGTH; index++) {
      if (Arrays.equals(stringOneContent, stringTwoContent)) {
        return true;
      }

      int charToAdd = s2.charAt(index) - 'a';
      int charToRemove = s2.charAt(index - S1_LENGTH) - 'a';

      stringTwoContent[charToAdd]++;
      stringTwoContent[charToRemove]--;
    }

    return Arrays.equals(stringOneContent, stringTwoContent);
  }
}
